### ISSUE TITLE
TS2729 false positive: assignment to optional property

### ISSUE THREAD

**Issue by Skalman** [microsoft/TypeScript#42789]
# Bug Report

### 🔎 Search Terms

TS2729, 2729, optional property use before initialization

### 🕗 Version & Regression Information

- This changed between versions 3.9.7 and 4.0.5
- The bug is also present in 4.2.0-dev.20210207

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.2.0-beta#code/MYGwhgzhAEDC0G8BQ1XTAfgFzQgFwCcBLAOwHMBuJFNAI2gF5Ea00ALAUxBAHscAKPGyIQAdGAAFTAEQB3HgRAATaQEoANC1bQA9DugAFAjwAOHAngCe0AORgb0EdACuEDkui0OAMwUdHeDCkRHhEYCBEAF5goTwkovwATADsiQCcqloAvlRZSEA)

### 💻 Code

```ts
class C {
    a?: string;

    b = {
        hello: (this.a = "world"), // <-- error
    };
}
```

<details>
<summary>Workaround</summary>

Neither of the following produces any errors.

```ts
class C {
    a!: string | undefined;

    b = {
        hello: (this.a = "world"),
    };
}

class C {
    a?: string = undefined;

    b = {
        hello: (this.a = "world"),
    };
}
```

</details>

### 🙁 Actual behavior

I get an error: `Property 'a' is used before its initialization.(2729)`

This is wrong for two reasons:
1. The property is _assigned_, not "used"
1. Since I used `?: string`, the property should be considered implicitly initialized to undefined

### 🙂 Expected behavior

No error

### Related issue

The same issue is described in #41516 "Google feedback on TS 4.0".
See the heading "Uninitialized fields used before being defined".
